doc/mtracebis.8.in
.\" This file was originally generated by help2man 1.44.1.
.TH MTRACEBIS "8" "February 2018" "mtracebis 0.1" "System Administration Utilities"
.SH NAME
mtracebis \- a program to initiate multicast traceroute "mtrace" queries
.SH SYNOPSIS
mtracebis
<multicast source>
.SH DESCRIPTION
.B mtracebis
is a program used to test multicast connectivity in a multicast and multicast
traceroute enabled IP network.
.PP
Initial version of the program requires multicast source IP address and
initiates a weak traceroute across the network. This tests whether the
interfaces towards the source are multicast enabled. First query sent is a
full query, capable of crossing the network all the way to the source. If this
fails, hop-by-hop queries are initiated.
.PP
Hop-by-hop queries start by requesting only a response from the nearest router.
Following that, next query is extended to the next two routers, and so on...
until a set of routers is tested for connectivity.
.SH SEE ALSO
See the project homepage at <@PACKAGE_URL@>.
.SH AUTHORS
Copyright 2018 Mladen Sablic


NEWS
Note: this file lists major user-visible changes only.

* Changes in Quagga 0.99.24

User-visible changes:
- [pimd] New daemon: pimd provides IPv4 PIM-SSM multicast routing.
- [bgpd] New feature: "next-hop-self all" to override nexthop on iBGP route
    reflector setups.
- [bgpd] route-maps have a new action "set ipv6 next-hop peer-address"
- [bgpd] route-maps have a new action "set as-path prepend last-as"
- [bgpd] Update validity checking (particularly MP-BGP / IPv6 routes) was
    touched up significantly.  Please report possible bugs.
- [ripd] New feature: RIP for IPv4 now supports equal-cost multipath (ECMP)
- [zebra] Multicast RIB support has been extended.  It still is IPv4 only.
- [zebra] "no link-detect" is now printed in configurations since it won't
    be the default anymore soon.  To retain current behaviour, re-save your
    configuration after updating to 0.99.24.

Distributor-visible changes:
- --enable-pimd is added to enable pimd.  It is considered experimental, though
    unless the distribution target is embedded systems with little flash, there
    is no reason to not include it in packages.
- --disable-ipv6 no longer exists as an option.  It's 2015, your C library
    really needs to have IPv6 support by now.
- --disable-netlink no longer exists as an option.  It didn't work anyway.
- --disable-solaris no longer exists as an option.  It only controlled some
    init scripts.
- --enable-isisd is now the default.
- mrlg.cgi is no longer included (it was severely outdated).  It can be found
    independently at http://mrlg.op-sec.us/
- build on Linux with the musl C library should now work

* Changes in Quagga 0.99.23

Known issues:
- [bgpd] setting an extcommunity in a route map on a route that already has
  an extcommunity attribute will cause bgpd to crash.  This issue will be
  fixed in a followup minor release.

User-visible changes:
- [lib] Performance enhancements on hashes and timers.
- [bgpd] New feature: iBGP TTL security.
- [bgpd] New feature: relaxed bestpath criteria for multipath and improved
  display of multipath routes in "show ip bgp".  Scripts parsing this output
  may need to be updated.
- [bgpd] Multiprotocol peerings over IPv6 now try to find a more appropriate
  IPv4 nexthop by looking at the interface.
- [ospf6d] A large amount of changes has been merged for ospf6d.  Careful
  evaluation prior to deployment is recommended.
- [zebra] Recursive route support has been overhauled.  Scripts parsing
  "show ip route" output may need adaptation.
- [zebra] IPv6 address management has been improved regarding tentative
  addresses.  This is visible in that a freshly configured address will not
  immediately be marked as usable.
- [*] a lot of bugs have been fixed, please refer to the git log

* Changes in Quagga 0.99.22

- [bgpd] The semantics of default-originate route-map have changed.
  The route-map is now used to advertise the default route conditionally.
  The old behaviour which allowed to set attributes on the originated
  default route is no longer supported.
- [bgpd] There is now a replace-as option to neighbor ... local-as ...
  no-prepend.  For details, refer to the user documentation.
- [zebra] An FPM interface has been added.  This provides an alternate
  interface to routing information and is geared at OpenFlow & co.
- [snmp] AgentX is now supported;  the old smux backend is considered
  deprecated.  ospf6d has also had OSPFV3-MIB added.
- [*] several issues with configuration save/load/apply have been fixed,
  in particular on ospf "max-metric router-lsa administrative" and
  "distribute-list", bgpd "no neighbor activate", isisd "metric-style",
- [*] a lot of bugs have been fixed, please refer to the git log

* Changes in Quagga 0.99.21

- [bgpd] BGP multipath support has been merged
- [bgpd] SAFI (Multicast topology) support has been extended to propagate
  the topology to zebra.
- [bgpd] AS path limit functionality has been removed
- [babeld] a new routing daemon implementing the BABEL ad-hoc mesh routing
  protocol has been merged.
- [isisd] a major overhaul has been picked up. Please note that isisd is
  STILL NOT SUITABLE FOR PRODUCTION USE.
- [*] a lot of bugs have been fixed, please refer to the git log

* Changes in Quagga 0.99.10

- [bgpd] 4-byte AS support added
- [bgpd] MRT format changes to version 2. Those relying on
  bgpd MRT table dumps may need to update their tools.
- [bgpd] Added new route-map set statement: "as-path exclude"
- Zebra RIB updates queue has evolved into a multi-level
  structure to address RIB consistency issues.

* Changes in Quagga 0.99.2

- [bgpd] Work queues added to bgpd to split up update processing,
  particularly beneficial when a peer session goes down. AS_PATH
  parsing rewritten to be clearer, more robust and ready for 4-byte.

- [ripd] Simple authentication is no longer the default authentication
  mode for ripd. The default is now no-authentication. Any setups which
  used simple authentication will probably need to update their
  configuration manually.

- [ospfd] 1s dead-interval with sub-second Hellos feature added.
  SPF timers now specified in milliseconds, and with adaptive
  hold-time support. RFC3137 Stub-router support added. Default ABR
  type is now 'cisco'.

- Solaris least privileges support added.

* Changes in Quagga 0.99.1

- Zserv is now buffered via threads and non-blocking in most cases for both
  clients and zebra, which should improve responsiveness of daemons when
  they must send many messages to zebra.

- 'show thread cpu' now displays both cpu+system and wall-clock time,
  where getrusage() is available.

- Background threads added and workqueue API added, with a
  'show work-queues' command. Thread scheduling improved slightly.

- Zebra now has a work-queue for RIB processing. See 'show work-queues' in
  the zebra daemon vty.

- Support for interface renaming on Linux netlink systems.

- GNU Zebra bgpd merges, including BGP Graceful-restart and "match ip

pimd/mtracebis.c
/*
 * Multicast Traceroute for FRRouting
 * Copyright (C) 2018  Mladen Sablic
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; see the file COPYING; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
 */

#ifdef __linux__

#include "pim_igmp_mtrace.h"

#include "checksum.h"
#include "mtracebis_routeget.h"

#include <sys/select.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/time.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <net/if.h>

#define MTRACEBIS_VERSION "0.1"
#define MTRACE_TIMEOUT (5)

#define IP_HDR_LEN (sizeof(struct ip))
#define IP_RA_LEN (4)
#define MTRACE_BUF_LEN (MTRACE_HDR_SIZE + (MTRACE_MAX_HOPS * MTRACE_RSP_SIZE))
#define IP_AND_MTRACE_BUF_LEN (IP_HDR_LEN + IP_RA_LEN + MTRACE_BUF_LEN)

static const char *progname;
static void usage(void)
{
	fprintf(stderr, "Usage : %s <multicast source>\n", progname);
}
static void version(void)
{
	fprintf(stderr, "%s %s\n", progname, MTRACEBIS_VERSION);
}

static int send_query(int fd, struct in_addr to_addr,
		      struct igmp_mtrace *mtrace)
{
	struct sockaddr_in to;
	socklen_t tolen;
	int sent;

	memset(&to, 0, sizeof(to));
	to.sin_family = AF_INET;
	to.sin_addr = to_addr;
	tolen = sizeof(to);

	sent = sendto(fd, (char *)mtrace, sizeof(*mtrace), MSG_DONTWAIT,
		      (struct sockaddr *)&to, tolen);

	if (sent < 1)
		return -1;
	return 0;
}

static void print_query(struct igmp_mtrace *mtrace)
{
	char src_str[INET_ADDRSTRLEN];
	char dst_str[INET_ADDRSTRLEN];
	char grp_str[INET_ADDRSTRLEN];

	printf("* Mtrace from %s to %s via group %s\n",
	       inet_ntop(AF_INET, &mtrace->src_addr, src_str, sizeof(src_str)),
	       inet_ntop(AF_INET, &mtrace->dst_addr, dst_str, sizeof(dst_str)),
	       inet_ntop(AF_INET, &mtrace->grp_addr, grp_str, sizeof(grp_str)));
}

static int recv_response(int fd, long msec, int *hops)
{
	int recvd;
	char mtrace_buf[IP_AND_MTRACE_BUF_LEN];
	struct ip *ip;
	struct igmp_mtrace *mtrace;
	int mtrace_len;
	int responses;
	int i;
	u_short sum;

	recvd = recvfrom(fd, mtrace_buf, IP_AND_MTRACE_BUF_LEN, 0, NULL, 0);

	if (recvd < 1) {
		fprintf(stderr, "recvfrom error: %s\n", strerror(errno));
		return -1;
	}

	if (recvd < (int)sizeof(struct ip)) {
		fprintf(stderr, "no ip header\n");
		return -1;
	}

	ip = (struct ip *)mtrace_buf;

	if (ip->ip_v != 4) {
		fprintf(stderr, "IP not version 4\n");
		return -1;
	}

	sum = ip->ip_sum;
	ip->ip_sum = 0;

	if (sum != in_cksum(ip, ip->ip_hl * 4))
		return -1;

	mtrace = (struct igmp_mtrace *)(mtrace_buf + (4 * ip->ip_hl));

	mtrace_len = ntohs(ip->ip_len) - ip->ip_hl * 4;

	if (mtrace_len < (int)MTRACE_HDR_SIZE)
		return -1;

	sum = mtrace->checksum;
	mtrace->checksum = 0;
	if (sum != in_cksum(mtrace, mtrace_len)) {
		fprintf(stderr, "mtrace checksum wrong\n");
		return -1;
	}

	if (mtrace->type != PIM_IGMP_MTRACE_RESPONSE)
		return -1;


	responses = mtrace_len - sizeof(struct igmp_mtrace);
	responses /= sizeof(struct igmp_mtrace_rsp);

	printf("%ld ms received responses from %d hops.\n", msec, responses);

	if (hops)
		*hops = responses;

	for (i = 0; i < responses; i++) {
		struct igmp_mtrace_rsp *rsp = &mtrace->rsp[i];

		if (rsp->fwd_code != 0)
			printf("-%d fwd. code 0x%2x.\n", i, rsp->fwd_code);
	}

	return 0;
}

static int wait_for_response(int fd, int *hops)
{
	fd_set readfds;
	struct timeval timeout;
	int ret = -1;
	long msec, rmsec, tmsec;

	FD_ZERO(&readfds);
	FD_SET(fd, &readfds);

	memset(&timeout, 0, sizeof(timeout));

	timeout.tv_sec = MTRACE_TIMEOUT;

	tmsec = timeout.tv_sec * 1000 + timeout.tv_usec / 1000;
	do {
		ret = select(fd + 1, &readfds, NULL, NULL, &timeout);
		if (ret <= 0)
			return ret;
		rmsec = timeout.tv_sec * 1000 + timeout.tv_usec / 1000;
		msec = tmsec - rmsec;
	} while (recv_response(fd, msec, hops) != 0);

	return ret;
}

int main(int argc, char *const argv[])
{
	struct in_addr mc_source;
	struct in_addr iface_addr;
	struct in_addr gw_addr;
	struct in_addr mtrace_addr;
	struct igmp_mtrace mtrace;
	int hops = 255;
	int rhops;
	int maxhops = 255;
	int perhop = 3;
	int ifindex;
	int unicast = 1;
	int ttl = 64;
	int fd = -1;
	int ret = -1;
	int c;
	int i, j;
	char ifname[IF_NAMESIZE];
	char ip_str[INET_ADDRSTRLEN];

	mtrace_addr.s_addr = inet_addr("224.0.1.32");

	uid_t uid = getuid();

	if (uid != 0) {
		printf("must run as root\n");
		exit(EXIT_FAILURE);
	}

	if (argc <= 0)
		progname = "mtracebis";
	else
		progname = argv[0];

	if (argc != 2) {
		usage();
		exit(EXIT_FAILURE);
	}

	while (1) {
		static struct option long_options[] = {
			{"help", no_argument, 0, 'h'},
			{"version", no_argument, 0, 'v'},
			{0, 0, 0, 0} };
		int option_index = 0;

		c = getopt_long(argc, argv, "vh", long_options, &option_index);

		if (c == -1)
			break;

		switch (c) {
		case 'h':
			usage();
			exit(0);
		case 'v':
			version();
			exit(0);
		default:
			usage();
			exit(EXIT_FAILURE);
		}
	}
	if (inet_pton(AF_INET, argv[1], &mc_source) != 1) {
		usage();
		fprintf(stderr, "%s: %s not a valid IPv4 address\n", argv[0],
			argv[1]);
		exit(EXIT_FAILURE);
	}

	ifindex = routeget(mc_source, &iface_addr, &gw_addr);
	if (ifindex < 0) {
		fprintf(stderr, "%s: failed to get route to source %s\n",
			argv[0], argv[1]);
		exit(EXIT_FAILURE);
	}

	if (if_indextoname(ifindex, ifname) == NULL) {
		fprintf(stderr, "%s: if_indextoname error: %s\n", argv[0],
			strerror(errno));
		exit(EXIT_FAILURE);
	}

	/* zero mtrace struct */
	memset((char *)&mtrace, 0, sizeof(mtrace));

	/* set up query */
	mtrace.type = PIM_IGMP_MTRACE_QUERY_REQUEST;
	mtrace.hops = hops;
	mtrace.checksum = 0;
	mtrace.grp_addr.s_addr = 0;
	mtrace.src_addr = mc_source;
	mtrace.dst_addr = iface_addr;
	mtrace.rsp_addr = unicast ? iface_addr : mtrace_addr;
	mtrace.rsp_ttl = ttl;
	mtrace.qry_id = 0xffffff & time(NULL);

	mtrace.checksum = in_cksum(&mtrace, sizeof(mtrace));

	fd = socket(AF_INET, SOCK_RAW, IPPROTO_IGMP);

	if (fd < 1) {
		fprintf(stderr, "%s: socket error: %s\n", argv[0],
			strerror(errno));
		exit(EXIT_FAILURE);
	}

	ret = setsockopt(fd, SOL_SOCKET, SO_BINDTODEVICE, ifname,
			 strlen(ifname));

	if (ret < 0) {
		fprintf(stderr, "%s: setsockopt error: %s\n", argv[0],
			strerror(errno));
		ret = EXIT_FAILURE;
		goto close_fd;
	}

	print_query(&mtrace);
	if (send_query(fd, gw_addr, &mtrace) < 0) {
		fprintf(stderr, "%s: sendto error: %s\n", argv[0],
			strerror(errno));
		ret = EXIT_FAILURE;
		goto close_fd;
	}
	printf("Querying full reverse path...\n");
	ret = wait_for_response(fd, NULL);
	if (ret > 0) {
		ret = 0;
		goto close_fd;
	}
	if (ret < 0) {
		fprintf(stderr, "%s: select error: %s\n", argv[0],
			strerror(errno));
		ret = EXIT_FAILURE;
		goto close_fd;
	}
	printf(" * ");
	printf("switching to hop-by-hop:\n");
	printf("%3d  ? (%s)\n", 0,
	       inet_ntop(AF_INET, &mtrace.dst_addr, ip_str, sizeof(ip_str)));
	for (i = 1; i < maxhops; i++) {
		printf("%3d ", -i);
		mtrace.hops = i;
		for (j = 0; j < perhop; j++) {
			mtrace.qry_id++;
			mtrace.checksum = 0;
			mtrace.checksum = in_cksum(&mtrace, sizeof(mtrace));
			if (send_query(fd, gw_addr, &mtrace) < 0) {
				fprintf(stderr, "%s: sendto error: %s\n",
					argv[0], strerror(errno));
				ret = EXIT_FAILURE;
				goto close_fd;
			}
			ret = wait_for_response(fd, &rhops);
			if (ret > 0) {
				if (i > rhops) {
					ret = 0;
					goto close_fd;
				}
				break;
			}
			printf(" *");
		}
		if (ret <= 0)
			printf("\n");
	}
	ret = 0;
close_fd:
	close(fd);
	exit(ret);
}

#else /* __linux__ */

#include <stdio.h>
#include <stdlib.h>

int main(int argc, char *argv[])
{
	printf("%s implemented only for GNU/Linux\n", argv[0]);
	exit(0);
}

#endif /* __linux__ */
